res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,884 @@
+                this.inSetValue(this.jogScale(value));
+            }
+        },
+    });
+
+    // Jog Wheel LED
+    this.jogWheelLed = new components.Component({
+        midi: [0x90 + midiChannel, 0x21],
+        on: colDeck[deckNumbers - 1],
+        trigger: function() {
+            this.send(this.on);
+        },
+    });
+
+    // Slip Mode Button
+    this.slipButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x24],
+        key: "slip_enabled",
+        type: components.Button.prototype.types.toggle,
+    });
+
+    // Loop In Button
+    this.loopInButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x25],
+        key: "loop_in",
+    });
+
+    // Loop Out Button
+    this.loopOutButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x26],
+        key: "loop_out",
+    });
+
+    // Load Buttons
+    this.deckLoad = new components.Button({
+        midi: [0x9F, midiChannel - 3],
+        key: "LoadSelectedTrack",
+
+        // TODO: Make load buttons reflect colour of currently active deck.
+        //       This is harder than expected due to the fact that I rely on the MIDI channel
+        //       of each physical deck component to determine which deck it's manipulating in
+        //       Mixxx. The deck load buttons, however, are both on MIDI channel 16, so I need
+        //       a better way to determine this.
+        off: Prime4.rgbCode.whiteDark,
+        on: Prime4.rgbCode.white,
+    });
+
+    this.reconnectComponents(function(c) {
+        if (c.group === undefined) {
+            c.group = this.currentDeck;
+        }
+    });
+};
+
+Prime4.Deck.prototype = new components.Deck();
+
+Prime4.shift = false;
+Prime4.shiftState = function(channel, control, value) {
+    Prime4.shift = value === 0x7F;
+    if (Prime4.shift) {
+        midi.sendShortMsg(0x90 + channel, control, 0x02);
+        Prime4.leftDeck.shift();
+        Prime4.rightDeck.shift();
+        Prime4.effectBank[0].shift();
+        Prime4.effectBank[1].shift();
+    } else {
+        midi.sendShortMsg(0x90 + channel, control, 0x01);
+        Prime4.leftDeck.unshift();
+        Prime4.rightDeck.unshift();
+        Prime4.effectBank[0].unshift();
+        Prime4.effectBank[1].unshift();
+    }
+};
+
+//========== PERFORMANCE PADS ==========//
+
+// Access the appropriate mode-select pad without remembering MIDI values
+Prime4.padMode = {
+    HOTCUE: 0x0B,
+    LOOP: 0x0C,
+    ROLL: 0x0D,
+    SLICER: 0x0E,
+};
+
+Prime4.PadSection = function(deck, offset) {
+    components.ComponentContainer.call(this);
+
+    // Create component containers for each pad mode
+    const modes = new components.ComponentContainer({
+        "hotcue": new Prime4.hotcueMode(deck, offset),
+        "loop": new Prime4.loopMode(deck, offset),
+        //TODO: "autoloop": new Prime4.autoloopMode(deck, offset),
+        "roll": new Prime4.rollMode(deck, offset),
+        "sampler": new Prime4.samplerMode(deck, offset),
+    });
```

Comment:
This might fix the fighting padmodes. Normally this shouldn't be necessary because we already specified `outConnect: false` in all the pads. 
```suggestion
    });
    modes.forEachComponent(c => c.disconnect());
```

Reply:
Aha! That did it! Now we just need to figure out how to update the pad mode selector button LEDs every time we switch decks

Reply:
There's also still the issue of the decks connecting to 3 and 4 on startup instead of 1 and 2

Reply:
> Aha! That did it! Now we just need to figure out how to update the pad mode selector button LEDs every time we switch decks

That shouldn't be difficult. Let me whip up a quick sketch. 

> There's also still the issue of the decks connecting to 3 and 4 on startup instead of 1 and 2

Can you clarify? Do you mean the input is routed to the wrong decks or the controller is reflecting the wrong output? With the current state of the code, it should still not reflect any LED feedback on any decks (ignore the mixer and other miscellaneous parts).

Reply:
It looks like on startup, the physical decks control deck 3 and 4, and the jog wheel colours update as such. However, the deck assignment buttons highlight decks 1 and 2 as if they are selected. If it makes things easier I can provide a video.

Reply:
That is quite weird. A video would really help here. Thank you. I'm sorry for the inconvenience.

Reply:
Not a problem at all, the behaviour is quite strange to explain. I'll upload a video later today.

Reply:
I'll upload a video once this `TypeError` stuff is sorted out, as I can't load the mapping till it's fixed :)

Reply:
Never mind, commented out the error-causing line. Here's the video - https://youtu.be/zfVBUfR72Q4